### ISSUE TITLE
Discussion: Use Redux?

### ISSUE THREAD

**Issue by D1no** [Rotorsoft/firebase-event-store#4]
I was wondering about the exact same thing you wrote in your opening paragraph. I was thinking to use REDUX on the server / cloud function to achieve the same result. I was wondering what you think about that and if you considered it before?

This is inspired from an „old“ talk on CQRS: 
[![Event Sourcing](http://img.youtube.com/vi/LrkXWj9zxsU/0.jpg)](http://www.youtube.com/watch?v=LrkXWj9zxsU "Event Sourcing")

At the end, around minute 23:30, he iterates again that redux is isomorphic for CQRS.

And this one „REDUX Actions on the Server“:
[![Event Sourcing Redux](http://img.youtube.com/vi/R_NENZ1AkvQ/0.jpg)](https://youtu.be/R_NENZ1AkvQ?t=10m20s "Event Sourcing Redux")

Anyway, great work!

**Comment by Rotorsoft** [comment]
Hi. I have never used redux on the server side, and on the client site I prefer the simpler combination of Vue and Vuex to store client state. Interesting videos though, kind of overcomplicated if you compare with the option to just listen to your projections via Firestore query listeners

**Comment by D1no** [comment]
Yeah I am kind of evaluating it as a means of using an existing knowledge domain (Redux) to achieve CQRS-like reliability. The last video introduces [swarm.js](https://github.com/gritzko/swarm) & and their [lodux](https://github.com/john202020/lodux/tree/v1) to share actions across all devices, including the server. The way in which they replicate their projections goes super complicated on the thing by running own websocket servers etc... something firebase simply replaces (i.e. with query listeners)

When I first came into the need to create an ES model on firebase (back than the real-time database) I was thinking to just dispatch actions into a collection and have firebase take care of syncing them (their websocket implementation). Cloud functions could than listen for actions and resolve them with another redux instance running in the cloud — making the redux store our projection playground. Still having all the actions a client ever issued as a log, being able to timetravel and replay them on the server.

Anyway, I‘m rambling. The need for my ES solution vaporized but the challenge stuck. My idea of bringing redux (or any other of the immutable state machines) to the server, was a double hit on 
- using a well tested and known system (instead of coding up a „new one“ like you did)
- homogenize the overall approach to data ingestion on the client and server (to me it always had been odd that we do CQRS on the server but still use upsert on the client)

tl;dr
To me, it seems, running redux on the server and using Firestore as the redux store looks to me like a [quick win in getting to a solid event sourcing](https://www.quora.com/Can-Redux-be-used-for-server-side-Node-js-without-any-UI) implementation.

If you have time at some point to take a brief look at redux, I‘d love to hear where you see the differences to the current CQRS implementation on Firestore.

**Comment by Rotorsoft** [comment]
I totally understand your points but didn’t want to miss the fun of coding something like this instead of marrying to another framework. It’s not a big deal though and any reliable logging system will do. 
About the second point I try to build my clients as read only reflections of the projections on my CQRS servers where all the business logic resides. I know there are scenarios where this might be a challenge but I haven’t hit that wall yet. To me the simplicity of a dumb client just responsible for presenting views and issuing commands is priceless and Vuex is completely capable of handling the reactive side of the GUI. I will take a second look at redux and comment in the future. Thanks for your thoughts!
